src/components/QueryBuilder/index.js

```diff
@@ -0,0 +1,298 @@
+import { Fragment, useEffect, useMemo, useState } from "react";
+import { Menu, Transition } from "@headlessui/react";
+import { MdDelete } from "react-icons/md";
+import { BiCog } from "react-icons/bi";
+import clsx from "clsx";
+import { useSearchParams } from "react-router-dom";
+import {
+  createSavedQuery,
+  deleteSavedQuery,
+  getAllSavedQueries,
+  getConfigsByQuery,
+  updateSavedQuery
+} from "../../api/services/configs";
+import { Modal } from "../Modal";
+import { toastError, toastSuccess } from "../Toast/toast";
+import { useLoader } from "../../hooks";
+import { TextInputClearable } from "../TextInputClearable";
+
+export const QueryBuilder = ({ refreshConfigs, className, ...props }) => {
+  const [params, setParams] = useSearchParams();
+  const [query, setQuery] = useState(params.get("query") || "");
+  const [selectedQuery, setSelectedQuery] = useState(null);
+  const [queryList, setQueryList] = useState([]);
+  const [modalIsOpen, setModalIsOpen] = useState(false);
+  const [savedQueryValue, setSavedQueryValue] = useState("");
+  const { loading, setLoading } = useLoader();
+  const options = useMemo(() => {
+    const data = [];
+
+    if (!selectedQuery) {
+      data.push({
+        id: "save",
+        label: "Save",
+        type: "action_save",
+        context: {}
+      });
+    }
+
+    if (selectedQuery) {
+      data.push({
+        id: "update",
+        label: "Update",
+        type: "action_update",
+        context: {}
+      });
+      data.push({
+        id: "save_as",
+        label: "Save As",
+        type: "action_save",
+        context: {}
+      });
+    }
+
+    queryList.forEach((queryItem) => {
+      data.push({
+        id: queryItem.id,
+        label: queryItem.description,
+        type: "action_saved_query",
+        context: {
+          ...queryItem
+        }
+      });
+    });
+    return data;
+  }, [queryList, selectedQuery]);
```

Comment:
I think the logic would be simpler if actions are rendered explicitly and their actions handled explicitly. Instead of having everything in array. Also, from UI perspective, it makes sense to separate action (save, save as, update) from saved queries (even if it's just a horizontal separator).

Reply:
https://github.com/flanksource/flanksource-ui/commit/4824a1ea2442e3e7e41c98189108f70760882ca6 is this more readable?

Reply:
@ciju Thanks for the suggestion, I am already doing the refactoring.